File: benchmarks/sources/js/binary_trees.js
/* The Great Computer Language Shootout
   http://shootout.alioth.debian.org/
   contributed by Isaac Gouy */

function TreeNode(left,right,item){
  this.left = left;
  this.right = right;
  this.item = item;
}

TreeNode.prototype.itemCheck = function(){
  if (this.left==null) return this.item;
  else return this.item + this.left.itemCheck() - this.right.itemCheck();
}

function bottomUpTree(item,depth){
  if (depth>0){
    return new TreeNode(
      bottomUpTree(2*item-1, depth-1)
      ,bottomUpTree(2*item, depth-1)
      ,item
    );
  }
  else {
    return new TreeNode(null,null,item);
  }
}


var minDepth = 4;
var n = 0; //arguments[0];
var maxDepth = Math.max(minDepth + 2, n);
var stretchDepth = maxDepth + 1;

var check = bottomUpTree(0,stretchDepth).itemCheck();
//print("stretch tree of depth " + stretchDepth + "\t check: " + check);

var longLivedTree = bottomUpTree(0,maxDepth);
for (var depth=minDepth; depth<=maxDepth; depth+=2){
  var iterations = 1 << (maxDepth - depth + minDepth);

  check = 0;
  for (var i=1; i<=iterations; i++){
    check += bottomUpTree(i,depth).itemCheck();
    check += bottomUpTree(-i,depth).itemCheck();
  }
  //   print(iterations*2 + "\t trees of depth " + depth + "\t check: " + check);
}

//print("long lived tree of depth " + maxDepth + "\t check: "
//   + longLivedTree.itemCheck());

File: benchmarks/sources/ml/binary_trees.ml
(* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * Contributed by Troestler Christophe
 * Modified by Fabrice Le Fessant
 *)

type 'a tree =
  | Empty
  | Node of 'a tree * 'a * 'a tree

let rec make i d =
  (* if d = 0 then Empty *)
  if d = 0
  then Node (Empty, i, Empty)
  else
    let i2 = 2 * i and d = d - 1 in
    Node (make (i2 - 1) d, i, make i2 d)

let rec check = function
  | Empty -> 0
  | Node (l, i, r) -> i + check l - check r

let min_depth = 4

let max_depth =
  let n = try int_of_string Sys.argv.(1) with _ -> 10 in
  max (min_depth + 2) n

let stretch_depth = max_depth + 1

let () =
  (* Gc.set { (Gc.get()) with Gc.minor_heap_size = 1024 * 1024; max_overhead = -1; }; *)
  let _c = check (make 0 stretch_depth) in
  ( (*
  Printf.printf "stretch tree of depth %i\t check: %i\n" stretch_depth c
 *) )

let long_lived_tree = make 0 max_depth

let loop_depths d =
  for i = 0 to ((max_depth - d) / 2) + 1 - 1 do
    let d = d + (i * 2) in
    let niter = 1 lsl (max_depth - d + min_depth) in
    let c = ref 0 in
    for i = 1 to niter do
      c := !c + check (make i d) + check (make (-i) d)
    done;
    ( (*
      Printf.printf "%i\t trees of depth %i\t check: %i\n" (2 * niter) d !c;
 *) )
  done

let () =
  (*
  flush stdout;
*)
  loop_depths min_depth;
  ( (*
  Printf.printf "long lived tree of depth %i\t check: %i\n"
    max_depth (check long_lived_tree)
 *) )
